### PR TITLE
Update transaction and enable OOO for regular transactions

### PR DESCRIPTION
**Pull request by dranikpg**

### CODE REVIEW

File: src/server/transaction.cc [hunk 1]
```diff
@@ -458,55 +457,41 @@ void Transaction::ScheduleInternal() {
...
     if (success.load(memory_order_acquire) == num_shards) {
-      // We allow out of order execution only for single hop transactions.
-      // It might be possible to do it for multi-hop transactions as well but currently is
-      // too complicated to reason about.
-      if (single_hop && lock_granted_cnt.load(memory_order_relaxed) == num_shards) {
-        // OOO can not happen with span-all transactions. We ensure it in ScheduleInShard when we
-        // refuse to acquire locks for these transactions..
-        DCHECK(!span_all);
+      coordinator_state_ |= COORD_SCHED;
+      if (!span_all && lock_granted_cnt.load(memory_order_relaxed) == num_shards) {
```
Comment (dranikpg) [inline src/server/transaction.cc:474]: This is the only functional change

File: src/server/journal/journal.cc [hunk 2]
```diff
@@ -87,16 +87,6 @@ void Journal::UnregisterOnChange(uint32_t id) {
   journal_slice.UnregisterOnChange(id);
 }
 
-bool Journal::SchedStartTx(TxId txid, unsigned num_keys, unsigned num_shards) {
```
Comment (romange) [inline src/server/journal/journal.cc:90]: unrelated. cleanups?

File: src/server/transaction.cc [hunk 3]
```diff
@@ -456,57 +455,45 @@ void Transaction::ScheduleInternal() {
...
+        // Currently we don't support OOO for incremental locking. Sp far they are global.
+        DCHECK(!(multi_ && multi_->is_expanding));
         coordinator_state_ |= COORD_OOO;
       }
       VLOG(2) << "Scheduled " << DebugId()
               << " OutOfOrder: " << bool(coordinator_state_ & COORD_OOO)
               << " num_shards: " << num_shards;
 
-      if (mode == IntentLock::EXCLUSIVE) {
```
Comment (romange) [inline src/server/transaction.cc:489]: i see now

File: src/server/transaction.cc [hunk 4]
```diff
@@ -549,75 +536,63 @@ void Transaction::MultiData::AddLocks(IntentLock::Mode mode) {
...
-  // single hop -> concluding.
-  coordinator_state_ |= (COORD_EXEC | COORD_EXEC_CONCLUDING);
+  DCHECK(multi_ || (coordinator_state_ & COORD_SCHED) == 0);   // Only multi schedule in advance.
+  coordinator_state_ |= (COORD_EXEC | COORD_EXEC_CONCLUDING);  // Single hop means we conclude.
 
-  if (!multi_) {  // for non-multi transactions we schedule exactly once.
-    DCHECK_EQ(0, coordinator_state_ & COORD_SCHED);
-  }
-
-  bool schedule_fast = (unique_shard_cnt_ == 1) && !IsGlobal() && !multi_;
```
Comment (romange) [inline src/server/transaction.cc:562]: can you please keep `schedule_fast` variable?

File: src/server/transaction.cc [hunk 5]
```diff
@@ -549,75 +536,63 @@ void Transaction::MultiData::AddLocks(IntentLock::Mode mode) {
...
-  bool run_eager = false;
+  bool was_ooo = false;
 
-  if (schedule_fast) {  // Single shard (local) optimization.
-    // We never resize shard_data because that would affect MULTI transaction correctness.
+  // If we run only on one shard and conclude, we can avoid scheduling at all
+  // and dispatch a single task to the its shard set.
```
Comment (romange) [inline src/server/transaction.cc:547]: `and dispatch a single task to the its shard set` -> `and directly dispatch the task to its destination shard`

File: src/server/transaction.cc [hunk 6]
```diff
@@ -549,75 +536,63 @@ void Transaction::MultiData::AddLocks(IntentLock::Mode mode) {
...
-    // since ScheduleSingleHop can finish before ScheduleUniqueShard returns.
-    // The problematic flow is as follows: ScheduleUniqueShard schedules into TxQueue
-    // (hence run_fast is false), and then calls PollExecute that in turn runs
-    // the callback which calls DecreaseRunCnt.
-    // As a result WaitForShardCallbacks below is unblocked before schedule_cb returns.
-    // However, if run_fast is true, then we may mutate stack variables, but only
-    // before DecreaseRunCnt is called.
-    auto schedule_cb = [&] {
+    // NOTE: schedule_cb cannot update data on stack when run_fast is false.
+    // This is because this function can finish before the callback returns.
```
Comment (romange) [inline src/server/transaction.cc:558]: `this function` - not clear which one you are referring to. Maybe just use `ScheduleSingleHop`

File: src/server/transaction.cc [hunk 7]
```diff
@@ -549,75 +536,63 @@ void Transaction::MultiData::AddLocks(IntentLock::Mode mode) {
...
-      if (multi_->is_expanding)
-        multi_->AddLocks(Mode());
-    } else {
+    // This transaction either spans multiple shards and/or is multi.
+
+    if (!multi_)  // Multi schedule in advance.
       ScheduleInternal();
-    }
+
+    if (multi_ && multi_->is_expanding)
```
Comment (romange) [inline src/server/transaction.cc:582]: nit: why did you change it? now we have 2 checks `if (!multi)`, `if (multi)` instead of a single check

**Review by romange — COMMENTED** [review]
looks good. minor comments

File: src/server/transaction.cc (hunk 7 above)
Comment (dranikpg) [reply src/server/transaction.cc:582]: It was a nested if before, which in my mind is harder to read, so I made it a single if-else if. And instead of this nested if, I just split it up. Now we have a easily visible condition->action part

**Review by romange — APPROVED** [review]
